### PR TITLE
[Perf][Kernel] stride-aware RMSNormKernel — drop contiguous() copy on aligned-non-contiguous inputs

### LINKED ISSUE #1065: [Perf][Kernel] stride-aware kernels to drop contiguous() copy on dim != -1
### Symptom / Motivation

Canonical-pattern ops migrating off the legacy `(M, N)` ctor (PRs #1061, #1064 and future scan / norm / reduction / softmax migrations) include the pipeline:

```python
if dim_norm != ndim - 1:
    x = x.movedim(dim_norm, -1)
post_move_shape = tuple(x.shape)
x = x.contiguous().reshape(-1, self.N)   # full read+write copy when dim != -1 or input non-contiguous
```

`x.contiguous()` triggers a full tensor copy whenever the input is non-contiguous along the inner dim — i.e., whenever the user passes `dim != -1` (default) or feeds a sliced view. For typical LLM workloads `dim = -1` is universal and the cost is zero, but it is still a known opt-in tax baked into every canonical-pattern op.

Surfaced by a gemini-code-assist review on PR #1064.

### Root Cause Analysis

Tilelang kernels (`RMSNormKernel`, `CumulativeKernel`, …) currently assume the input is row-major contiguous `(M, N)` 2-D. When the manifest declares `dim` as a `signature.params` entry (canonical pattern per `docs/ops-design.md` § Step 3), the op layer is the only place that can normalize layout — hence the host-side `movedim + contiguous + reshape`.

The principled fix is at the kernel ABI: pass strides through and let the kernel issue strided loads, removing the host-side copy.

### Related Files

- `tileops/ops/norm/norm_base.py` — `RowNormOp._validate_and_normalize_dim` / `_flatten_to_2d`
- `tileops/ops/reduction/cumulative_base.py` — `CumulativeOp._run`
- `tileops/kernels/norm/rms_norm.py` — first prototype target (smallest kernel, single dispatch, well-tested)
- `tileops/kernels/reduction/cumulative.py` — second target after RMS proves the pattern

## Goal

Selected row-norm / scan / reduction kernels accept `(x, M, N, stride_M, stride_N)` and use `T.copy` with explicit strides instead of assuming row-major contiguous layout. Op-layer canonical-pattern bases drop the `x.contiguous()` call when feeding these kernels and pass the strides through verbatim. Net effect: zero host-side copy on `dim != -1` workloads, no regression on `dim = -1`.

## Plan

**Plan type: proposal**

1. Prototype on `RMSNormKernel`: extend ctor / build to accept `stride_M, stride_N` (default `(N, 1)` for backwards-compat), thread through to `T.copy`. Keep existing call sites unchanged.
2. Add a stride-aware path in `RowNormOp._run`: when `dim != -1`, skip `contiguous()`, pass `x.stride(dim_norm), x.stride(...)` to the kernel.
3. Measure perf on a representative `dim != -1` workload (e.g., RMS over a non-last axis on a Llama-style 3-D shape). Compare against the contiguous-copy baseline.
4. Decide GO / NO-GO based on the measurement: GO → roll out the same pattern to `CumulativeKernel` (`CumulativeOp._run`) and any future canonical-pattern kernels; NO-GO → revert the kernel change, document the cost as accepted.
5. If GO: file a follow-up issue listing every canonical-pattern kernel that needs the stride-aware ABI.

## Constraints

- **Inner-dim contiguity (`stride_N == 1`) is non-negotiable** for coalesced GPU loads. `movedim(dim, -1)` only changes which axis is inner, not the inner axis's contiguity, so this is fine — but the kernel ABI must enforce it (assert / static guard).
- **Kernel-layer only**. This issue tracks the kernel ABI change. The op-layer canonical pattern (PRs #1061, #1064) ships as-is — no rollback there.
- **No regression on `dim = -1`**. The default path must be free (`stride_M = N, stride_N = 1` is the contiguous case; tile-level codegen should compile to the same loads).

## Acceptance Criteria

- [ ] AC-1: Modified files pass existing unit tests.
- [ ] AC-2: `RMSNormKernel` accepts optional `stride_M, stride_N` ctor args; existing call sites with no stride args produce identical output (bit-exact on `dim = -1` workloads).
- [ ] AC-3: A new test in `tests/ops/test_rms_norm.py` exercises a `dim != -1` shape (e.g., 3-D `(B, H, S)` with `dim = 1`) using the stride-aware path and matches the PyTorch reference within the existing `atol`.
- [ ] AC-4: Measured perf delta on a representative `dim != -1` workload is reported in the PR (e.g., `(2, 4096, 512)` bf16 with `dim = 1` on H200), comparing the new stride-aware path vs. the host-side `contiguous()` baseline.
- [ ] AC-5: `RowNormOp._run` either drops `x.contiguous()` (GO path) or keeps it with a documented reason (NO-GO path).
- [ ] AC-6: Decision (GO / NO-GO) and follow-up scope (which kernels to extend) recorded in the PR description.

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1065

## Summary

Decision: **GO**.

Prototype the stride-aware kernel ABI on `RMSNormKernel` so `RMSNormFwdOp` can drop `x.contiguous()` when the input is aligned-non-contiguous. The kernel now declares its global input as `T.StridedTensor((M, N_padded), (stride_M, 1), dtype)` and accepts an optional `stride_M` ctor arg; default `stride_M = N_padded` keeps the contiguous codegen bit-identical to the prior `T.Tensor` form, so no regression on `dim = -1`.

`RMSNormFwdOp.forward` calls a new `_try_strided_2d_view` helper that builds a zero-copy 2-D `as_strided` view when leading dims are row-major-contiguous AND the inner stride is 1; on that path `x.contiguous()` is skipped and the per-row stride flows through to the kernel. When the layout is incompatible (e.g. `movedim(1, -1)` on a 3-D `(B,H,S)` input — leading dims become non-contiguous), the op falls back to the original `contiguous()` path.

## Scope

Kernel + Op layer for RMS only. Tests added for the new strided path. Other canonical-pattern kernels (LayerNorm, fused-add norms, GroupNorm, InstanceNorm, AdaLayerNorm{,Zero}, BatchNorm, CumulativeKernel) follow the same migration template once this lands — tracked as a follow-up issue per the plan.

## Test plan

- [x] **AC-1**: 18/18 `pytest tests/ops/test_rms_norm.py` passes
- [x] **AC-2**: `RMSNormKernel` accepts optional `stride_M`; default value reproduces bit-identical output vs the explicit `stride_M = N_padded` form on `dim=-1` workloads (verified by direct probe)
- [x] **AC-3**: New `test_rms_norm_strided_zero_copy` exercises sliced 3-D `(2, 4, 4096)` from `(2, 4, 8192)` for fp16/bf16 — strided path matches `torch.nn.functional.rms_norm` reference within existing tolerances
- [x] **AC-4**: Perf measured on H200 (untuned)

  | dtype | shape | strided | baseline (`contiguous()`+kernel) | speedup |
  |---|---|---|---|---|
  | fp16 | B=2 H=4 N=4096 (of 8192) | 0.061 ms | 0.069 ms | 1.12x |
  | fp16 | B=4 H=8 N=4096 (of 8192) | 0.052 ms | 0.061 ms | 1.19x |
  | bf16 | B=2 H=4 N=4096 (of 8192) | 0.055 ms | 0.070 ms | 1.28x |
  | bf16 | B=4 H=8 N=4096 (of 8192) | 0.048 ms | 0.099 ms | 2.06x |

  Contiguous-path baseline (no regression vs prior code, codegen bit-identical): fp16 M=1024 N=4096 → 0.053 ms; M=8192 N=8192 → 0.084 ms

- [x] **AC-5**: `RMSNormFwdOp.forward` drops `x.contiguous()` on the GO path; falls back when layout is incompatible (documented in helper docstring)
- [x] **AC-6**: GO decision recorded above; follow-up scope = LayerNorm / fused-add norms / GroupNorm / InstanceNorm / AdaLayerNorm{,Zero} / BatchNorm / CumulativeKernel

## Constraints honored

- **`stride_N == 1` non-negotiable**: the kernel raises `ValueError` when the inner stride isn't 1, so callers can't accidentally produce non-coalesced loads
- **No regression on `dim = -1`**: default `stride_M = N_padded` reuses the contiguous codegen — verified bit-identical
- **Kernel-layer prototype only**: the issue tracks the kernel ABI change; this PR ships kernel + op-layer caller change + a single targeted test, no manifest or workload edits

## Limitation (deferred)

The literal `(B, H, S)` with `dim=1` example in the issue, after `movedim(1, -1)`, has strides `(H*S, 1, S)` — inner stride `S != 1` violates the kernel's coalescing constraint. That layout cannot fit a 2-D `StridedTensor` and falls back to `contiguous()` in this PR. A 3-D-aware kernel or a true-`dim != -1` reduction kernel is outside this prototype's scope and belongs in a follow-up.

### CODE REVIEW

File: tileops/kernels/norm/rms_norm.py [hunk 1]
```diff
@@ -20,22 +25,28 @@
...
 
 @functools.lru_cache(maxsize=32)
-def _rms_norm_kernel(M, N, eps, dtype):
+def _rms_norm_kernel(M, N, eps, dtype, stride_M):
     N_padded = _align_up(N, ALIGNMENT)
+    if stride_M < N_padded:
+        raise ValueError(
+            f"stride_M ({stride_M}) must be >= N_padded ({N_padded}); "
+            "inner dim must be contiguous (stride_N == 1)"
+        )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/norm/rms_norm.py:42]: Adopted. Reworded the error to describe the row-overlap invariant directly; the inner-stride constraint is already enforced by the constant in T.StridedTensor. See 7ce631f.

File: tileops/ops/norm/rms_norm.py [hunk 2]
```diff
@@ -15,6 +15,31 @@
...
+    if x.shape[-1] != n:
+        return None
+    inner_stride = x.stride(-2) if x.ndim >= 2 else 1
+    # Check leading-dim contiguity: stride[i] == size[i+1] * stride[i+1].
+    for i in range(x.ndim - 2):
+        if x.stride(i) != x.shape[i + 1] * x.stride(i + 1):
+            return None
+    m = 1
+    for d in x.shape[:-1]:
+        m *= int(d)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/rms_norm.py]: Adopted. Switched to math.prod. See 7ce631f.
